File: web/lib/grin-header.ts
import {
  HEADER_FIELDS,
  HEADER_SIZE_BYTES,
  MAGIC_BYTES,
  MAX_RULE_COUNT,
  PIXEL_SIZE_BYTES,
  RULES_BLOCK_SIZE,
  VERSION_MAJOR,
  VERSION_MINOR,
} from "./format";

export type GrinValidationResult = {
  ok: boolean;
  errors: string[];
  warnings: string[];
};

export type GrinHeaderInit = {
  width: number;
  height: number;
  tickMicros?: number;
  ruleCount?: number;
  opcodeSetId?: number;
  flags?: number;
  pixelDataLength?: bigint;
  fileLength?: bigint;
  pixelDataOffset?: bigint;
  reservedA?: bigint;
  reservedB?: bigint;
  rulesBlock?: Uint8Array;
  versionMajor?: number;
  versionMinor?: number;
  magic?: Uint8Array;
  headerSize?: number;
};

export class GrinHeader {
  magic: Uint8Array;
  versionMajor: number;
  versionMinor: number;
  headerSize: number;
  width: number;
  height: number;
  tickMicros: number;
  ruleCount: number;
  opcodeSetId: number;
  flags: number;
  pixelDataLength: bigint;
  fileLength: bigint;
  pixelDataOffset: bigint;
  reservedA: bigint;
  reservedB: bigint;
  rulesBlock: Uint8Array;

  constructor(init: GrinHeaderInit) {
    this.magic = copyBytes(init.magic ?? MAGIC_BYTES, HEADER_FIELDS.MAGIC.size);
    this.versionMajor = init.versionMajor ?? VERSION_MAJOR;
    this.versionMinor = init.versionMinor ?? VERSION_MINOR;
    this.headerSize = init.headerSize ?? HEADER_SIZE_BYTES;
    this.width = init.width;
    this.height = init.height;
    this.tickMicros = init.tickMicros ?? 0;
    this.ruleCount = init.ruleCount ?? 0;
    this.opcodeSetId = init.opcodeSetId ?? 0;
    this.flags = init.flags ?? 0;
    const expectedPixelLength =
      BigInt(this.width) * BigInt(this.height) * BigInt(PIXEL_SIZE_BYTES);
    this.pixelDataLength = init.pixelDataLength ?? expectedPixelLength;
    this.fileLength = init.fileLength ?? 0n;
    this.pixelDataOffset = init.pixelDataOffset ?? BigInt(HEADER_SIZE_BYTES);
    this.reservedA = init.reservedA ?? 0n;
    this.reservedB = init.reservedB ?? 0n;
    this.rulesBlock = copyBytes(
      init.rulesBlock ?? new Uint8Array(RULES_BLOCK_SIZE),
      RULES_BLOCK_SIZE
    );
  }

  static deserialize(bytes: Uint8Array): GrinHeader {
    if (bytes.length < HEADER_SIZE_BYTES) {
      throw new Error(`Header requires ${HEADER_SIZE_BYTES} bytes`);
    }
    const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    const magic = copyBytes(
      bytes.subarray(HEADER_FIELDS.MAGIC.offset, HEADER_FIELDS.MAGIC.offset + HEADER_FIELDS.MAGIC.size),
      HEADER_FIELDS.MAGIC.size
    );
    const versionMajor = view.getUint8(HEADER_FIELDS.VERSION_MAJOR.offset);
    const versionMinor = view.getUint8(HEADER_FIELDS.VERSION_MINOR.offset);
    const headerSize = view.getUint16(HEADER_FIELDS.HEADER_SIZE.offset, true);
    const width = view.getUint32(HEADER_FIELDS.WIDTH.offset, true);
    const height = view.getUint32(HEADER_FIELDS.HEIGHT.offset, true);
    const tickMicros = view.getUint32(HEADER_FIELDS.TICK_MICROS.offset, true);
    const ruleCount = view.getUint8(HEADER_FIELDS.RULE_COUNT.offset);
    const opcodeSetId = view.getUint8(HEADER_FIELDS.OPCODE_SET_ID.offset);
    const flags = view.getUint16(HEADER_FIELDS.FLAGS.offset, true);
    const pixelDataLength = view.getBigUint64(HEADER_FIELDS.PIXEL_DATA_LENGTH.offset, true);
    const fileLength = view.getBigUint64(HEADER_FIELDS.FILE_LENGTH.offset, true);
    const pixelDataOffset = view.getBigUint64(HEADER_FIELDS.PIXEL_DATA_OFFSET.offset, true);
    const reservedA = view.getBigUint64(HEADER_FIELDS.RESERVED_A.offset, true);
    const reservedB = view.getBigUint64(HEADER_FIELDS.RESERVED_B.offset, true);
    const rulesBlock = copyBytes(
      bytes.subarray(
        HEADER_FIELDS.RULES_BLOCK.offset,
        HEADER_FIELDS.RULES_BLOCK.offset + HEADER_FIELDS.RULES_BLOCK.size
      ),
      RULES_BLOCK_SIZE
    );

    return new GrinHeader({
      magic,
      versionMajor,
      versionMinor,
      headerSize,
      width,
      height,
      tickMicros,
      ruleCount,
      opcodeSetId,
      flags,
      pixelDataLength,
      fileLength,
      pixelDataOffset,
      reservedA,
      reservedB,
      rulesBlock,
    });
  }

  serialize(): Uint8Array {
    const buffer = new ArrayBuffer(HEADER_SIZE_BYTES);
    const view = new DataView(buffer);
    const bytes = new Uint8Array(buffer);

    bytes.set(this.magic, HEADER_FIELDS.MAGIC.offset);
    view.setUint8(HEADER_FIELDS.VERSION_MAJOR.offset, this.versionMajor);
    view.setUint8(HEADER_FIELDS.VERSION_MINOR.offset, this.versionMinor);
    view.setUint16(HEADER_FIELDS.HEADER_SIZE.offset, this.headerSize, true);
    view.setUint32(HEADER_FIELDS.WIDTH.offset, this.width, true);
    view.setUint32(HEADER_FIELDS.HEIGHT.offset, this.height, true);
    view.setUint32(HEADER_FIELDS.TICK_MICROS.offset, this.tickMicros, true);
    view.setUint8(HEADER_FIELDS.RULE_COUNT.offset, this.ruleCount);
    view.setUint8(HEADER_FIELDS.OPCODE_SET_ID.offset, this.opcodeSetId);
    view.setUint16(HEADER_FIELDS.FLAGS.offset, this.flags, true);
    view.setBigUint64(HEADER_FIELDS.PIXEL_DATA_LENGTH.offset, this.pixelDataLength, true);
    view.setBigUint64(HEADER_FIELDS.FILE_LENGTH.offset, this.fileLength, true);
    view.setBigUint64(HEADER_FIELDS.PIXEL_DATA_OFFSET.offset, this.pixelDataOffset, true);
    view.setBigUint64(HEADER_FIELDS.RESERVED_A.offset, this.reservedA, true);
    view.setBigUint64(HEADER_FIELDS.RESERVED_B.offset, this.reservedB, true);
    bytes.set(this.rulesBlock, HEADER_FIELDS.RULES_BLOCK.offset);

    return bytes;
  }

  validate(): GrinValidationResult {
    const errors: string[] = [];
    const warnings: string[] = [];

    if (!matchesMagic(this.magic)) {
      errors.push("Magic bytes do not match GRIN");
    }

    if (this.headerSize !== HEADER_SIZE_BYTES) {
      errors.push(`HeaderSize must be ${HEADER_SIZE_BYTES}`);
    }

    if (this.ruleCount > MAX_RULE_COUNT) {
      errors.push(`RuleCount exceeds ${MAX_RULE_COUNT}`);
    }

    if (this.pixelDataOffset !== BigInt(HEADER_SIZE_BYTES)) {
      errors.push("PixelDataOffset64 must be 128");
    }

    const expectedPixelLength =
      BigInt(this.width) * BigInt(this.height) * BigInt(PIXEL_SIZE_BYTES);
    if (this.pixelDataLength !== expectedPixelLength) {
      errors.push("PixelDataLength does not match width * height * 5");
    }

    const minFileLength = BigInt(HEADER_SIZE_BYTES) + this.pixelDataLength;
    if (this.fileLength !== 0n && this.fileLength < minFileLength) {
      errors.push("FileLength is smaller than header + pixel data");
    }

    if (this.rulesBlock.length !== RULES_BLOCK_SIZE) {
      errors.push("RulesBlock must be 64 bytes");
    }

    if (this.flags !== 0) {
      warnings.push("Flags field is non-zero");
    }

    if (this.reservedA !== 0n || this.reservedB !== 0n) {
      warnings.push("Reserved header fields are non-zero");
    }

    if (this.versionMajor !== VERSION_MAJOR || this.versionMinor !== VERSION_MINOR) {
      warnings.push("Unexpected GRIN version");
    }

    return { ok: errors.length === 0, errors, warnings };
  }

  clone(overrides: Partial<GrinHeaderInit> = {}): GrinHeader {
    return new GrinHeader({
      magic: overrides.magic ?? this.magic,
      versionMajor: overrides.versionMajor ?? this.versionMajor,
      versionMinor: overrides.versionMinor ?? this.versionMinor,
      headerSize: overrides.headerSize ?? this.headerSize,
      width: overrides.width ?? this.width,
      height: overrides.height ?? this.height,
      tickMicros: overrides.tickMicros ?? this.tickMicros,
      ruleCount: overrides.ruleCount ?? this.ruleCount,
      opcodeSetId: overrides.opcodeSetId ?? this.opcodeSetId,
      flags: overrides.flags ?? this.flags,
      pixelDataLength: overrides.pixelDataLength ?? this.pixelDataLength,
      fileLength: overrides.fileLength ?? this.fileLength,
      pixelDataOffset: overrides.pixelDataOffset ?? this.pixelDataOffset,
      reservedA: overrides.reservedA ?? this.reservedA,
      reservedB: overrides.reservedB ?? this.reservedB,
      rulesBlock: overrides.rulesBlock ?? this.rulesBlock,
    });
  }
}

function matchesMagic(bytes: Uint8Array): boolean {
  if (bytes.length !== MAGIC_BYTES.length) {
    return false;
  }
  for (let i = 0; i < MAGIC_BYTES.length; i += 1) {
    if (bytes[i] !== MAGIC_BYTES[i]) {
      return false;
    }
  }
  return true;
}

function copyBytes(source: Uint8Array, size: number): Uint8Array {
  if (source.length !== size) {
    const copy = new Uint8Array(size);
    copy.set(source.subarray(0, size));
    return copy;
  }
  return new Uint8Array(source);
}

File: android/lib/src/main/kotlin/io/grin/lib/GrinHeader.kt
package io.grin.lib

import java.nio.ByteBuffer
import java.nio.ByteOrder

class GrinHeader(
    magic: ByteArray,
    val versionMajor: Int,
    val versionMinor: Int,
    val headerSize: Int,
    val width: Long,
    val height: Long,
    val tickMicros: Long,
    val ruleCount: Int,
    val opcodeSetId: Int,
    val flags: Int,
    val pixelDataLength: Long,
    val fileLength: Long,
    val pixelDataOffset: Long,
    val reservedA: Long,
    val reservedB: Long,
    rulesBlock: ByteArray
) {
    val magic: ByteArray = magic.copyOf()
    val rulesBlock: ByteArray = rulesBlock.copyOf()

    init {
        require(this.magic.size == GrinHeaderLayout.MAGIC_SIZE) {
            "Magic bytes must be ${GrinHeaderLayout.MAGIC_SIZE} bytes"
        }
        require(this.rulesBlock.size == GrinFormat.RULES_BLOCK_SIZE) {
            "RulesBlock must be ${GrinFormat.RULES_BLOCK_SIZE} bytes"
        }
    }

    fun serialize(): ByteArray {
        val buffer = ByteBuffer.allocate(GrinFormat.HEADER_SIZE_BYTES).order(ByteOrder.LITTLE_ENDIAN)
        putBytes(buffer, GrinHeaderLayout.MAGIC_OFFSET, magic)
        putUInt8(buffer, GrinHeaderLayout.VERSION_MAJOR_OFFSET, versionMajor)
        putUInt8(buffer, GrinHeaderLayout.VERSION_MINOR_OFFSET, versionMinor)
        putUInt16(buffer, GrinHeaderLayout.HEADER_SIZE_OFFSET, headerSize)
        putUInt32(buffer, GrinHeaderLayout.WIDTH_OFFSET, width)
        putUInt32(buffer, GrinHeaderLayout.HEIGHT_OFFSET, height)
        putUInt32(buffer, GrinHeaderLayout.TICK_MICROS_OFFSET, tickMicros)
        putUInt8(buffer, GrinHeaderLayout.RULE_COUNT_OFFSET, ruleCount)
        putUInt8(buffer, GrinHeaderLayout.OPCODE_SET_ID_OFFSET, opcodeSetId)
        putUInt16(buffer, GrinHeaderLayout.FLAGS_OFFSET, flags)
        putUInt64(buffer, GrinHeaderLayout.PIXEL_DATA_LENGTH_OFFSET, pixelDataLength)
        putUInt64(buffer, GrinHeaderLayout.FILE_LENGTH_OFFSET, fileLength)
        putUInt64(buffer, GrinHeaderLayout.PIXEL_DATA_OFFSET_OFFSET, pixelDataOffset)
        putUInt64(buffer, GrinHeaderLayout.RESERVED_A_OFFSET, reservedA)
        putUInt64(buffer, GrinHeaderLayout.RESERVED_B_OFFSET, reservedB)
        putBytes(buffer, GrinHeaderLayout.RULES_BLOCK_OFFSET, rulesBlock)
        return buffer.array()
    }

    fun validate(): GrinValidationResult {
        val errors = mutableListOf<String>()
        val warnings = mutableListOf<String>()

        if (!magic.contentEquals(GrinFormat.MAGIC_BYTES)) {
            errors.add("Magic bytes do not match GRIN")
        }
        if (headerSize != GrinFormat.HEADER_SIZE_BYTES) {
            errors.add("HeaderSize must be ${GrinFormat.HEADER_SIZE_BYTES}")
        }
        if (ruleCount > GrinFormat.MAX_RULE_COUNT) {
            errors.add("RuleCount exceeds ${GrinFormat.MAX_RULE_COUNT}")
        }
        if (pixelDataOffset != GrinFormat.HEADER_SIZE_BYTES.toLong()) {
            errors.add("PixelDataOffset64 must be 128")
        }

        val expectedPixelLength = expectedPixelLength(width, height)
        if (expectedPixelLength == null) {
            errors.add("PixelDataLength overflow for width/height")
        } else if (pixelDataLength != expectedPixelLength) {
            errors.add("PixelDataLength does not match width * height * 5")
        }

        val minFileLength = GrinFormat.HEADER_SIZE_BYTES.toLong() + pixelDataLength
        if (fileLength != 0L && fileLength < minFileLength) {
            errors.add("FileLength is smaller than header + pixel data")
        }

        if (flags != 0) {
            warnings.add("Flags field is non-zero")
        }
        if (reservedA != 0L || reservedB != 0L) {
            warnings.add("Reserved header fields are non-zero")
        }
        if (versionMajor != GrinFormat.VERSION_MAJOR || versionMinor != GrinFormat.VERSION_MINOR) {
            warnings.add("Unexpected GRIN version")
        }

        return GrinValidationResult(errors.isEmpty(), errors, warnings)
    }

    fun copy(
        magic: ByteArray = this.magic,
        versionMajor: Int = this.versionMajor,
        versionMinor: Int = this.versionMinor,
        headerSize: Int = this.headerSize,
        width: Long = this.width,
        height: Long = this.height,
        tickMicros: Long = this.tickMicros,
        ruleCount: Int = this.ruleCount,
        opcodeSetId: Int = this.opcodeSetId,
        flags: Int = this.flags,
        pixelDataLength: Long = this.pixelDataLength,
        fileLength: Long = this.fileLength,
        pixelDataOffset: Long = this.pixelDataOffset,
        reservedA: Long = this.reservedA,
        reservedB: Long = this.reservedB,
        rulesBlock: ByteArray = this.rulesBlock
    ): GrinHeader {
        return GrinHeader(
            magic = magic,
            versionMajor = versionMajor,
            versionMinor = versionMinor,
            headerSize = headerSize,
            width = width,
            height = height,
            tickMicros = tickMicros,
            ruleCount = ruleCount,
            opcodeSetId = opcodeSetId,
            flags = flags,
            pixelDataLength = pixelDataLength,
            fileLength = fileLength,
            pixelDataOffset = pixelDataOffset,
            reservedA = reservedA,
            reservedB = reservedB,
            rulesBlock = rulesBlock
        )
    }

    companion object {
        fun deserialize(bytes: ByteArray): GrinHeader {
            require(bytes.size >= GrinFormat.HEADER_SIZE_BYTES) {
                "Header requires ${GrinFormat.HEADER_SIZE_BYTES} bytes"
            }
            val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)

            val magic = bytes.copyOfRange(
                GrinHeaderLayout.MAGIC_OFFSET,
                GrinHeaderLayout.MAGIC_OFFSET + GrinHeaderLayout.MAGIC_SIZE
            )
            val versionMajor = readUInt8(buffer, GrinHeaderLayout.VERSION_MAJOR_OFFSET)
            val versionMinor = readUInt8(buffer, GrinHeaderLayout.VERSION_MINOR_OFFSET)
            val headerSize = readUInt16(buffer, GrinHeaderLayout.HEADER_SIZE_OFFSET)
            val width = readUInt32(buffer, GrinHeaderLayout.WIDTH_OFFSET)
            val height = readUInt32(buffer, GrinHeaderLayout.HEIGHT_OFFSET)
            val tickMicros = readUInt32(buffer, GrinHeaderLayout.TICK_MICROS_OFFSET)
            val ruleCount = readUInt8(buffer, GrinHeaderLayout.RULE_COUNT_OFFSET)
            val opcodeSetId = readUInt8(buffer, GrinHeaderLayout.OPCODE_SET_ID_OFFSET)
            val flags = readUInt16(buffer, GrinHeaderLayout.FLAGS_OFFSET)
            val pixelDataLength = readUInt64(buffer, GrinHeaderLayout.PIXEL_DATA_LENGTH_OFFSET)
            val fileLength = readUInt64(buffer, GrinHeaderLayout.FILE_LENGTH_OFFSET)
            val pixelDataOffset = readUInt64(buffer, GrinHeaderLayout.PIXEL_DATA_OFFSET_OFFSET)
            val reservedA = readUInt64(buffer, GrinHeaderLayout.RESERVED_A_OFFSET)
            val reservedB = readUInt64(buffer, GrinHeaderLayout.RESERVED_B_OFFSET)
            val rulesBlock = bytes.copyOfRange(
                GrinHeaderLayout.RULES_BLOCK_OFFSET,
                GrinHeaderLayout.RULES_BLOCK_OFFSET + GrinFormat.RULES_BLOCK_SIZE
            )

            return GrinHeader(
                magic = magic,
                versionMajor = versionMajor,
                versionMinor = versionMinor,
                headerSize = headerSize,
                width = width,
                height = height,
                tickMicros = tickMicros,
                ruleCount = ruleCount,
                opcodeSetId = opcodeSetId,
                flags = flags,
                pixelDataLength = pixelDataLength,
                fileLength = fileLength,
                pixelDataOffset = pixelDataOffset,
                reservedA = reservedA,
                reservedB = reservedB,
                rulesBlock = rulesBlock
            )
        }
    }
}

private fun expectedPixelLength(width: Long, height: Long): Long? {
    if (width < 0L || height < 0L) {
        return null
    }
    val pixelCount = width * height
    if (width != 0L && pixelCount / width != height) {
        return null
    }
    val pixelLength = pixelCount * GrinFormat.PIXEL_SIZE_BYTES.toLong()
    if (pixelCount != 0L && pixelLength / pixelCount != GrinFormat.PIXEL_SIZE_BYTES.toLong()) {
        return null
    }
    return pixelLength
}

private fun readUInt8(buffer: ByteBuffer, offset: Int): Int {
    return buffer.get(offset).toInt() and 0xFF
}

private fun readUInt16(buffer: ByteBuffer, offset: Int): Int {
    return buffer.getShort(offset).toInt() and 0xFFFF
}

private fun readUInt32(buffer: ByteBuffer, offset: Int): Long {
    return buffer.getInt(offset).toLong() and 0xFFFFFFFFL
}

private fun readUInt64(buffer: ByteBuffer, offset: Int): Long {
    return buffer.getLong(offset)
}

private fun putUInt8(buffer: ByteBuffer, offset: Int, value: Int) {
    buffer.put(offset, (value and 0xFF).toByte())
}

private fun putUInt16(buffer: ByteBuffer, offset: Int, value: Int) {
    buffer.putShort(offset, (value and 0xFFFF).toShort())
}

private fun putUInt32(buffer: ByteBuffer, offset: Int, value: Long) {
    buffer.putInt(offset, (value and 0xFFFFFFFFL).toInt())
}

private fun putUInt64(buffer: ByteBuffer, offset: Int, value: Long) {
    buffer.putLong(offset, value)
}

private fun putBytes(buffer: ByteBuffer, offset: Int, bytes: ByteArray) {
    var index = 0
    while (index < bytes.size) {
        buffer.put(offset + index, bytes[index])
        index += 1
    }
}
